local/node.go

```diff
@@ -38,27 +42,126 @@ type getConnFunc func(context.Context, node.Node) (net.Conn, error)
 // AvalancheGo binaries in tests
 type NodeProcess interface {
 	// Start this process
-	Start() error
+	// Returns error if not called after instantiation
+	// If process stops without a previous call to Stop(), send notification msg over given channel
+	Start(chan network.UnexpectedNodeStopMsg) error
 	// Send a SIGTERM to this process
-	Stop() error
+	// If ctx is cancelled, send SIGKILL to this process and descendants
+	// Returns error if called before Start
+	// Returns nil if the process is already stopping/stopped
+	Stop(ctx context.Context) error
 	// Returns when the process finishes exiting
+	// Returns error if called before Start
+	// Returns nil if already waited for the process
 	Wait() error
+	// Returns true if the process is executing
+	// Returns false if the process has been stopped
+	// Returns false if the process has not been started
+	Alive() bool
 }
 
 type nodeProcessImpl struct {
-	cmd *exec.Cmd
+	name string
+	lock sync.RWMutex
+	cmd  *exec.Cmd
+	// to notify Wait() on process stop, and give wait return
+	waitReturnCh chan error
+	// to notify user of not asked process stops
+	unexpectedStopCh chan network.UnexpectedNodeStopMsg
+	// maintains process state Initial/Started/Stopping/Stopped/Waited
+	state int
+	// to notify SIGKILL goroutine of process end
+	closeOnStop chan struct{}
 }
 
-func (p *nodeProcessImpl) Start() error {
-	return p.cmd.Start()
+const (
+	Initial = iota
+	Started
+	Stopping
+	Stopped
+	Waited
+)
+
+// to be called only on Initial state
+func (p *nodeProcessImpl) Start(unexpectedStopCh chan network.UnexpectedNodeStopMsg) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	if p.state != Initial {
+		return errors.New("start called on invalid state")
+	}
+	p.unexpectedStopCh = unexpectedStopCh
+	if err := p.cmd.Start(); err != nil {
+		return err
+	}
+	p.state = Started
+	p.waitReturnCh = make(chan error, 1)
+	p.closeOnStop = make(chan struct{})
+	go func() {
+		// Wait4 to avoid race conditions on stdout/stderr pipes
+		var status syscall.WaitStatus
+		var rusage syscall.Rusage
+		_, err := syscall.Wait4(p.cmd.Process.Pid, &status, 0, &rusage)
+		p.waitReturnCh <- err
+		p.lock.Lock()
+		state := p.state
+		p.state = Stopped
+		close(p.closeOnStop)
+		p.lock.Unlock()
+		if state != Stopping {
+			p.unexpectedStopCh <- network.UnexpectedNodeStopMsg{
+				Name:     p.name,
+				ExitCode: status.ExitStatus(),
+			}
+		}
+	}()
```

Comment:
Why don't we just use `p.cmd.Wait()` to wait for the program to exit?

Reply:
In general I'd prefer we avoid using raw syscalls unless absolutely necessary

Reply:
I agree, this looks like overly complex, we should avoid doing this low-level stuff.

Reply:
Really can use Wait, but there is a race condition in a unit test. Wait closes process pipes and invalidates a unit test case. TestChildCmdRedirection see comment on StdoutPipe, StderrPipe. To avoid that just use Wait4. It seems to me at this point that or the test case is removed or Wait4 is retained.